source/common/http/utility.cc

```diff
@@ -751,6 +751,23 @@ const std::string& Utility::getProtocolString(const Protocol protocol) {
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
+const std::string Utility::buildOriginalUri(const Http::RequestHeaderMap& request_headers,
+                                            const uint32_t max_path_length) {
+  if (!request_headers.Path()) {
+    return "";
+  }
+  absl::string_view path(request_headers.EnvoyOriginalPath()
+                             ? request_headers.getEnvoyOriginalPathValue()
+                             : request_headers.getPathValue());
+
+  if (path.length() > max_path_length) {
+    path = path.substr(0, max_path_length);
+  }
+
+  return absl::StrCat(request_headers.getForwardedProtoValue(), "://",
```

Comment:
does ForwardedProto and Host always exist?

Reply:
It appears from this line 
https://github.com/envoyproxy/envoy/blob/64739c567af14f14895d24d2389f2206b0c35ef9/source/common/http/conn_manager_utility.cc#L147-L151

That the ForwardedProto is always set.

Host must be set and valid for a request to get to the filter, we can see that here:
https://github.com/envoyproxy/envoy/blob/64739c567af14f14895d24d2389f2206b0c35ef9/source/common/http/conn_manager_impl.cc#L945-L946